File: 0981 time-based-key-value-store/time-based-key-value-store.cpp
struct Pair {
    string value;
    int time;
    Pair(string v, int t) : value(v), time(t) {}
};

class TimeMap {
public:
    unordered_map<string, vector<Pair>> map;

    TimeMap() {}

    void set(string key, string value, int timestamp) {
        map[key].push_back(Pair(value, timestamp));
    }

    string get(string key, int timestamp) {
        if (map.find(key) == map.end()) return "";
        vector<Pair>& nums = map[key];
        int l = 0, r = nums.size() - 1;
        string ans = "";
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (nums[mid].time <= timestamp) {
                ans = nums[mid].value;
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return ans;
    }
};

File: 0981 time-based-key-value-store/time-based-key-value-store.java
class Pair {
    String value;
    int time;
    Pair(String value,int time) {
        this.value = value;
        this.time = time;
    }
}

class TimeMap {
    Map<String,List<Pair>> map;
    public TimeMap() {
        map = new HashMap<>();
    }
    
    public void set(String key, String value, int timestamp) {
        if(!map.containsKey(key)) {
            map.put(key,new ArrayList<>());
        }
        map.get(key).add(new Pair(value, timestamp));
    }
    
    public String get(String key, int timestamp) {
        if(!map.containsKey(key)) return "";
        List<Pair> nums = map.get(key);
        int l = 0,r = nums.size()-1;
        String ans = "";
        while(l<=r) {
            int mid = l+(r-l)/2;
            if(nums.get(mid).time <= timestamp) {
                ans = nums.get(mid).value;
                l=mid+1;
            } else {
                r=mid-1;
            }
        }
        return ans;
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap obj = new TimeMap();
 * obj.set(key,value,timestamp);
 * String param_2 = obj.get(key,timestamp);
 */